packages/search-input/src/SearchInput/SearchInput.tsx

```diff
@@ -0,0 +1,377 @@
+            focused: index === highlightIndex,
+          });
+        } else if (isComponentType(child, 'SearchResultGroup')) {
+          const nestedChildren = React.Children.map(
+            child.props.children,
+            processChild,
+          );
+
+          if (nestedChildren && nestedChildren.length > 0) {
+            return React.cloneElement(child, {
+              ...child.props,
+              children: nestedChildren,
+            });
+          }
+        }
+      };
+
+      const updatedChildren = React.Children.map(children, processChild);
+
+      return {
+        resultsCount,
+        updatedChildren,
+      };
+    };
+
+    const { updatedChildren, resultsCount } = processChildren();
+
+    type Direction = 'next' | 'prev' | 'first' | 'last';
+    const updateHighlight = (direction: Direction) => {
+      switch (direction) {
+        case 'first': {
+          setHighlightIndex(0);
+          break;
+        }
+
+        case 'last': {
+          setHighlightIndex(resultsCount);
+          break;
+        }
+
+        case 'next': {
+          const nextIndex =
+            !isUndefined(highlightIndex) && highlightIndex + 1 < resultsCount
+              ? highlightIndex + 1
+              : 0;
+          setHighlightIndex(nextIndex);
+          break;
+        }
+
+        case 'prev': {
+          const nextIndex =
+            !isUndefined(highlightIndex) && highlightIndex - 1 >= 0
+              ? highlightIndex - 1
+              : resultsCount - 1;
+          setHighlightIndex(nextIndex);
+        }
+      }
+    };
+
+    const isElementInComponent = (el: Element | null) => {
+      return formRef.current?.contains(el) || menuRef.current?.contains(el);
+    };
+
+    /** Event Handlers */
+
+    const handleOpenMenuAction: EventHandler<SyntheticEvent<any>> = e => {
+      if (disabled) {
+        e.preventDefault();
+        e.stopPropagation();
+      } else {
+        openMenu();
+      }
+    };
+
+    const handleSearchBoxMousedown = (e: React.MouseEvent) => {
+      if (disabled) {
+        // Prevent container from gaining focus by default
+        e.preventDefault();
+      }
+    };
+
+    const handleSearchBoxClick: MouseEventHandler = handleOpenMenuAction;
+
+    // Fired whenever the wrapper gains focus,
+    // and any time the focus within changes
+    const handleSearchBoxFocus: FocusEventHandler = e => {
+      const target =
+        e.target !== clearButtonRef.current
+          ? inputRef.current ?? (e.target as HTMLElement)
+          : clearButtonRef.current;
+      target.focus();
+      trackFocusedElement(target);
+      handleOpenMenuAction(e);
+    };
+
+    const handleClearButton: MouseEventHandler<HTMLButtonElement> = e => {
+      // We convert the click event into a Change event,
+      // Update the target & value,
+      // Then fire any `onChange` handler
+      const changeEvent: ChangeEvent<any> = {
+        ...cloneDeep(e),
```

Comment:
Why do we need to clone the event?

Reply:
In order to get the proper types & event properties. We can't call `handleChange(e)` since `e` is a `MouseEvent`

Reply:
Actually, we don't need to clone, we can just cast & change properties. Fixed this